server/src/test/java/org/elasticsearch/index/codec/tsdb/ES87TSDBDocValuesEncoderTests.java

```diff
@@ -260,12 +260,53 @@ public void testEncodeOrdinalsNoRepetitions() throws IOException {
         doTestOrdinals(arr, 113);
     }
 
+    public void testEncodeOrdinalsBitPack3Bits() throws IOException {
+        long[] arr = new long[blockSize];
+        Arrays.fill(arr, 4);
+        for (int i = 0; i < 4; i++) {
+            arr[i] = i;
+        }
+        doTestOrdinals(arr, 49);
+    }
+
+    public void testEncodeOrdinalsCycle() throws IOException {
+        long[] arr = new long[blockSize];
+        Arrays.setAll(arr, i -> i % 3);
+        doTestOrdinals(arr, 4);
+    }
+
+    public void testEncodeOrdinalsLongCycle() throws IOException {
+        long[] arr = new long[blockSize];
+        Arrays.setAll(arr, i -> i % 63);
+        doTestOrdinals(arr, 65);
+    }
+
+    public void testEncodeOrdinalsCycleTooLong() throws IOException {
+        long[] arr = new long[blockSize];
+        Arrays.setAll(arr, i -> i % 64);
+        // the cycle is too long and the vales are bit-packed
+        doTestOrdinals(arr, 97);
+    }
+
+    public void testEncodeOrdinalsAlmostCycle() throws IOException {
+        long[] arr = new long[blockSize];
+        Arrays.setAll(arr, i -> i % 3);
+        arr[arr.length -1] = 4;
+        doTestOrdinals(arr, 49);
+    }
+
+    public void testEncodeOrdinalsDifferentCycles() throws IOException {
+        long[] arr = new long[blockSize];
+        Arrays.setAll(arr, i -> i > 64 ? i % 4 : i % 3);
+        doTestOrdinals(arr, 33);
+    }
+
     private void doTestOrdinals(long[] arr, long expectedNumBytes) throws IOException {
         long maxOrd = 0;
         for (long ord : arr) {
             maxOrd = Math.max(maxOrd, ord);
         }
-        final int bitsPerOrd = PackedInts.bitsRequired(maxOrd - 1);
+        final int bitsPerOrd = PackedInts.bitsRequired(maxOrd);
```

Comment:
I noticed tests would fail that have values up to 4.
I'm not sure why I did `maxOrd -1` before. I think I copied it from places like this that @jpountz added:

https://github.com/elastic/elasticsearch/blob/f26691f9876d23e289357c30ba11314fa35aa5d1/server/src/main/java/org/elasticsearch/index/codec/tsdb/ES87TSDBDocValuesConsumer.java#L147

Do we need to adjust that as well? What's the reason you did `PackedInts.bitsRequired(maxOrd - 1)` vs `PackedInts.bitsRequired(maxOrd)`?